lib/browser-pool/caching-pool.js

```diff
@@ -35,7 +36,13 @@ CachingPool.prototype.getBrowser = function(id) {
     }
 
     log('has cached browser %o', browser);
-    return browser.reset().thenResolve(browser);
+    return browser.reset()
+        .fail(function(e) {
+            var reject = q.reject.bind(null, e);
+            return this.freeBrowser(browser)
```

Comment:
But probably you are right. Let this be the preflight test: if we can't reset browser session it probably broken, so discard this session and open new one.